src/plugins/gs-plugin-external-apps.c

```diff
@@ -0,0 +1,712 @@
+			     const gchar *repo_dir,
+			     const char *runtime,
+			     const char *asset,
+			     GError **error)
+{
+	g_autofree gchar *content_type = NULL;
+	g_autoptr(GFile) download_file = NULL;
+
+	char *download_name = gs_plugin_download_asset (plugin, app, runtime,
+							asset, error);
+
+	if (!download_name) {
+		return FALSE;
+	}
+
+	download_file = g_file_new_for_path (download_name);
+	content_type = gs_utils_get_content_type (download_file, NULL,
+						  error);
+
+	if (!content_type) {
+		return FALSE;
+	}
+
+	g_debug ("Adding runtime asset with type '%s'", content_type);
+
+	if (g_strcmp0 (content_type, "application/x-deb") == 0) {
+		gs_plugin_add_runtime_deb_asset (build_dir, repo_dir,
+						 download_name, error);
+	} else {
+		g_debug ("Cannot deal with asset type '%s'", content_type);
+
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+gs_plugin_build_runtime (GsPlugin *plugin,
+			 GsApp *app,
+			 GsApp *runtime,
+			 GError **error)
+{
+	gboolean ret;
+	g_autofree char *tmp_dir;
+	const char *template = "3rdparty_XXXXXX";
+	const char *build_dir = "3rd_party_repo";
+	g_autofree char *repo_dir = NULL;
+	g_autofree char *tmp_build_dir = NULL;
+	g_autofree char *repo_name = NULL;
+	g_autofree char *dir_basename = NULL;
+	const char *runtime_name = gs_app_get_id_no_prefix (runtime);
+	const char *runtime_url = gs_app_get_metadata_item (runtime,
+							    METADATA_URL);
+
+	tmp_dir = g_dir_make_tmp (template, error);
+
+	gs_app_set_progress (app, 5);
+
+	if (!tmp_dir) {
+		g_debug ("Could not create temporary dir for building a "
+			 "runtime extension!");
+		return FALSE;
+	}
+
+	g_debug ("Building runtime extension '%s' in dir '%s'", runtime_name,
+		 tmp_dir);
+
+	tmp_build_dir = g_strconcat (tmp_dir, "/", build_dir, NULL);
+	if (g_mkdir_with_parents (tmp_build_dir, 0777) != 0) {
+		g_debug ("Failed to make dir %s", tmp_build_dir);
+		return FALSE;
+	}
+
+	gs_app_set_progress (app, 10);
+
+	flatpak_repo_build_init (tmp_build_dir, runtime_name, error);
+
+	gs_app_set_progress (app, 15);
+
+	gs_plugin_add_runtime_asset (plugin, app, tmp_dir, tmp_build_dir,
+				     runtime_name, runtime_url, error);
+
+	repo_dir = g_strconcat (tmp_dir, "/", runtime_name, NULL);
+
+	gs_app_set_progress (app, 30);
+
+	g_debug ("Exporting repo in '%s'... (this may take a while)",
+		 tmp_build_dir);
+	ret = flatpak_repo_build_export (tmp_build_dir, repo_dir,
+					 error);
+
+	gs_app_set_progress (app, 35);
+
+	if (!ret) {
+		g_debug ("Failed to export repo '%s' in '%s'!",
+			 repo_dir, tmp_build_dir);
+		if (*error) {
+			g_debug ("Error: %s", (*error)->message);
+		}
+		return FALSE;
+	}
+
+	g_debug ("Repo '%s' exported!", repo_dir);
+
+	/* Use the temp dir random name as a prefix for the runtime. This avoids
+	 * any clashes with existing repos and makes cleaning up easier. */
+	dir_basename = g_path_get_basename (tmp_dir);
+	repo_name = g_strconcat (dir_basename, runtime_name, NULL);
+
+	gs_app_set_progress (app, 50);
```

Comment:
This will basically happen immediately after the previous call that sets the progress to `35`, since there's no real operation that happens in between...